File: cpp/94. Binary Tree Inorder Traversal.cpp
/*
2016-9-28 01:27:50

https://leetcode.com/problems/binary-tree-inorder-traversal/

Given a binary tree, return the inorder traversal of its nodes' values.

For example:
Given binary tree [1,null,2,3],
   1
    \
     2
    /
   3
return [1,3,2].

Note: Recursive solution is trivial, could you do it iteratively?
*/



/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
		stack<TreeNode*> s;
		TreeNode* iter = root;
		TreeNode* tmp;
		
		while ((iter != nullptr) || !s.empty())
		{
			if (iter != nullptr)
			{
				s.push(iter);
				iter = iter->left;
			}
			else
			{
				iter = s.top();
				s.pop();
				
				res.push_back(iter->val);
				
				iter = iter->right;
				
			}
		}
		
		return res;
    }
};

File: java/src/main/java/club/cser/leetcode/94. Binary Tree Inorder Traversal.java
package club.cser.leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BinaryTreeInorderTraversal {

    public List<Integer> inorderTraversal(TreeNode root) {

        Stack<TreeNode> path = new Stack<>();
        List<Integer> res = new ArrayList<>();
        TreeNode iter = root;

        while (null != iter || !path.empty()) {

            if (null != iter) {
                path.push(iter);
                iter = iter.left;
            } else {
                res.add(path.peek().val);
                iter = path.peek().right;
                path.pop();
            }

        }
        return res;
    }

    public List<Integer> inorderMorrisTraversal(TreeNode root) {


        List<Integer> res = new ArrayList<>();
        TreeNode cur = root;

        while (null != cur) {
            if (null != cur.left) {
                TreeNode rightMostNodeOfLeftSubtree = cur.left;

                while (rightMostNodeOfLeftSubtree.right != null && rightMostNodeOfLeftSubtree.right != cur)
                    rightMostNodeOfLeftSubtree = rightMostNodeOfLeftSubtree.right;

                if (rightMostNodeOfLeftSubtree.right == null) {
                    rightMostNodeOfLeftSubtree.right = cur;
                    cur = cur.left;
                } else {
                    rightMostNodeOfLeftSubtree.right = null;

                    res.add(cur.val);
                    cur = cur.right;
                }
            } else {
                res.add(cur.val);
                cur = cur.right;
            }
        }

        return res;
    }

}